kernel/src/hil/gpio.rs

```diff
@@ -1,67 +1,289 @@
-    /// be configured as an interrupt. The `identifier`
-    /// can be any value and will be returned to you
-    /// when the interrupt on this pin fires.
-    fn enable_interrupt(&self, identifier: usize, mode: InterruptMode);
+    /// Disable interrupts for the GPIO pin.
+    fn disable_interrupts(&self);
 
-    /// Disable the interrupt for the GPIO pin.
-    fn disable_interrupt(&self);
+    /// Return whether this interrupt is pending
+    fn is_pending(&self) -> bool;
 }
 
-/// Interface for users of synchronous GPIO. In order
+/// Interface for users of synchronous GPIO interrupts. In order
 /// to receive interrupts, the user must implement
 /// this `Client` interface.
 pub trait Client {
     /// Called when an interrupt occurs. The `identifier` will
     /// be the same value that was passed to `enable_interrupt()`
     /// when the interrupt was configured.
-    fn fired(&self, identifier: usize);
+    fn fired(&self);
+}
+
+/// Interface that wraps an interrupt to pass a value when it
+/// triggers. The standard use case for this trait is when several
+/// interrupts call the same callback function and it needs to
+/// distinguish which one is calling it by giving each one a unique
+/// value.
+pub trait InterruptWithValue: Input {
+    /// Set the client for interrupt events.
+    fn set_client(&self, client: &'static ClientWithValue);
+
+    /// Set the underlying interrupt source.
+    fn set_source(&'static self, source: &'static InterruptPin);
+
+    /// Enable an interrupt on the GPIO pin. This does not
+    /// configure the pin except to enable an interrupt: it
+    /// should be separately configured as an input, etc.
+    /// Returns:
+    ///    SUCCESS - the interrupt was set up properly
+    ///    FAIL    - the interrupt was not set up properly; this is due to
+    ///              not having an underlying interrupt source yet, i.e.
+    ///              the struct is not yet fully initialized.
+    fn enable_interrupts(&self, mode: InterruptEdge) -> ReturnCode;
+
+    /// Disable interrupts for the GPIO pin.
+    fn disable_interrupts(&self);
+
+    /// Return whether this interrupt is pending
+    fn is_pending(&self) -> bool;
+
+    /// Set the value that will be passed to clients on an
+    /// interrupt.
+    fn set_value(&self, value: u32);
+
+    /// Return the value that is passed to clients on an
+    /// interrupt.
+    fn value(&self) -> u32;
+}
+
+/// Interfaces for users of GPIO interrupts who handle many interrupts
+/// with the same function. The value passed in the callback allows the
+/// callback to distinguish which interrupt fired.
+pub trait ClientWithValue {
+    fn fired(&self, value: u32);
+}
+
+/// Standard implementation of InterruptWithValue: handles an
+/// `gpio::Client::fired` and passes it up as a
+/// `gpio::ClientWithValue::fired`.
+pub struct InterruptValueWrapper {
+    value: Cell<u32>,
+    client: OptionalCell<&'static ClientWithValue>,
```

Comment:
Would be nice to use `dyn` in all use of `trait objects` because trait objects without explicit `dyn` is deprecated in later rust versions 

For example `client: OptionalCell<&'static dyn ClientWithValue> `

Reply:
This seems like something we should do consistently over the entire repository. It seems like a good thing to propose for 1.4.